pkg/aaq-controller/built-in-usage-calculators/metadata-filter-calc.go

```diff
@@ -0,0 +1,65 @@
+package built_in_usage_calculators
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/kubernetes/pkg/quota/v1/evaluator/core"
+	"k8s.io/utils/clock"
+	"kubevirt.io/application-aware-quota/staging/src/kubevirt.io/application-aware-quota-api/pkg/apis/core/v1alpha1"
+)
+
+type MetadataFilterCalculator struct {
+	Labels      map[string]string
+	Annotations map[string]string
+	MyConfigs   []v1alpha1.VmiCalcConfigName
+	CalcConfig  v1alpha1.VmiCalcConfigName
+}
+
+func NewMetadataFilterCalculator(labels, annotations map[string]string, myConfigs []v1alpha1.VmiCalcConfigName, calcConfig v1alpha1.VmiCalcConfigName) *MetadataFilterCalculator {
+	return &MetadataFilterCalculator{
+		Labels:      labels,
+		Annotations: annotations,
+		MyConfigs:   myConfigs,
+		CalcConfig:  calcConfig,
+	}
+}
+
+var cdiLabels = map[string]string{"app": "containerized-data-importer"}
+
+var cdiConfigs = []v1alpha1.VmiCalcConfigName{v1alpha1.VirtualResources, v1alpha1.DedicatedVirtualResources}
+
+func NewCDIFilterCalculator(calcConfig v1alpha1.VmiCalcConfigName) *MetadataFilterCalculator {
+	return &MetadataFilterCalculator{
+		Labels:     cdiLabels,
+		MyConfigs:  cdiConfigs,
+		CalcConfig: calcConfig,
+	}
+}
+
+func (filtercalc *MetadataFilterCalculator) PodUsageFunc(pod *corev1.Pod, _ []*corev1.Pod) (corev1.ResourceList, error, bool) {
+	if !hasMatchingKeyAndValue(filtercalc.Labels, pod.Labels) &&
+		!hasMatchingKeyAndValue(filtercalc.Annotations, pod.Annotations) {
+		return corev1.ResourceList{}, nil, false
+	}
```

Comment:
Thanks for the PR! The approach of adding an additional calculator for CDI pods looks great. I do have some concerns about relying on labels to identify CDI pods. While we could apply the same method for virt-launcher pods, filtering based on labels might allow users to create pods with specific labels to bypass the quota.

Do you think it would be possible to verify that these pods are real CDI pods by checking through the VM/VMI objects, to prevent this?

Reply:
Hey @Barakmor1, I understand your concern but but I'm not sure going with owner references is a winning battle against a motivated user. CDI pods are owned by PVCs, not any CDI specific resource. And in the case of cross-namespace cloning, the source pods don't even have owner references because the "owner" may be in another namespace. So a user could get around the quota by creating a PVC + Pod that is owned by the PVC.

I think the only way to be 100% secure would be for HCO or come other component to publish a whitelist of allowed images+shas

Reply:
Thank you for the explanation, I'm not too familiar with how CDI pods work, but what we really need is a way to ensure that we have a VMI that justifies the existence of the pod we're evaluating and that there's only one relevant pod. 
Are you familiar with any approach to achieve this without HCO?

FYI @vladikr

Reply:
>but what we really need is a way to ensure that we have a VMI

Okay this is something I did not consider as a requirement. For each CDI pod, there is a "target" PVC. Could try to map that PVC to a VM/VMI that refers to it.

But what to do about standalone DVs/CDI populators? For example when managing "golden images"

Reply:
> Okay this is something I did not consider as a requirement. For each CDI pod, there is a "target" PVC. Could try to map that PVC to a VM/VMI that refers to it.

This approach sounds good to me.

> But what to do about standalone DVs/CDI populators? For example when managing "golden images"

As for the standalone DVs/CDI populators, I'm not sure we need to address this with the `VirtualResources` or `dedicatedVirtualResources` configurations. The goal here as i see it is to hide the infrastructure cost of VMIs, and if a VMI doesn’t exist, I don’t think we should change the resource counting. Users who want to ignore the pods related to these golden images can either use a dedicated namespace for them or apply their own pluggable policy for such cases.